File: PascalsTriangle.hs
-- an original implementation of Pascal's triangle
-- by Gage Swenson (gageswenson@gmail.com)
--
-- sponsored by ItLit
-- https://itlit.io
-- 

module Main where

import System.Environment (getArgs)

triangle :: Int -> [[Int]]
triangle 0 = [[1]]
triangle 1 = triangle 0 ++ [[1, 1]]
triangle n = 
    let top = triangle (n-1)
    in top ++ [makeRow 0 (last top)]
    where
    makeRow :: Int -> [Int] -> [Int]
    makeRow x rowAbove
        | x == 0    = 1 : makeRow (x+1) rowAbove
        | x == n    = [1]
        | otherwise = (rowAbove!!x + rowAbove!!(x-1)) : makeRow (x+1) rowAbove
        

showTriangle :: [[Int]] -> String
showTriangle [] = ""
showTriangle (r:rs) = replicate (length rs) '\t' ++ tabOut r ++ "\n" ++ showTriangle rs
    where
    tabOut :: [Int] -> String
    tabOut [c] = show c
    tabOut (c:cs) = show c ++ "\t\t" ++ tabOut cs

pascals :: Int -> String
pascals = showTriangle . triangle

main :: IO ()
main = do
    args <- getArgs
    case args of
        []     -> putStrLn "argc must be greater than 0"
        (x:xs) -> case isInt x of
            False -> putStrLn "argv[0] must be a positive int"
            True  -> putStr $ pascals (read x :: Int)
    where
    isInt :: String -> Bool
    isInt "" = False
    isInt [d]
        | isDigit d = True
        | otherwise = False
    isInt (d:ds)
        | isDigit d = True && isInt ds
        | otherwise = False
    isDigit :: Char -> Bool
    isDigit = (`elem` "1234567890")

File: PascalsTriangle.c
// Gage Swenson
// 03 July 2022

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

bool isInt(const char *d) {
    for (int i = 0; i < strlen(d); i++) {
        switch(d[i]) {
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9':
                continue;
        }
        return false;
    }
    return true;
}

int **makeTriangle(const int d) {
    int **triangle = malloc(sizeof(int*) * (d+1));
    triangle[0] = malloc(sizeof(int));
    triangle[0][0] = 1;
    if (d == 0)
        return triangle;
    triangle[1] = malloc(sizeof(int) * 2);
    triangle[1][0] = 1;
    triangle[1][1] = 1;
    for (int i = 2; i < d+1; i++) {
        triangle[i] = malloc(sizeof(int) * (i+1));
        triangle[i][0] = 1;
        triangle[i][i] = 1;
        for (int j = 1; j < i; j++)
            triangle[i][j] = triangle[i-1][j-1] + triangle[i-1][j];
    }
    return triangle;
}

void showTriangle(int **triangle, const int d) {
    for (int i = 0; i < d+1; i++) {
        for (int j = 0; j < d-i; j++)
            printf("\t");
        for (int j = 0; j < i; j++)
            printf("%d\t\t", triangle[i][j]);
        printf("%d\n", triangle[i][i]);
    }
}

void freeTriangle(int **triangle, const int d) {
    for (int i = 0; i < d+1; i++)
        free(triangle[i]);
    free(triangle);
}

int main(int argc, char **argv) {
    if (argc < 2) {
        printf("argc must be greater than 1\n");
        return 1;
    }
    const char *argv1 = argv[1];
    if (!isInt(argv1)) {
        printf("argv[1] must be a positive int\n");
        return 2;
    }
    const int d = atoi(argv1);
    int **triangle = makeTriangle(d);
    showTriangle(triangle, d);
    freeTriangle(triangle, d);
    return 0;
}
